relay-cardinality/src/redis.rs

```diff
@@ -341,6 +418,250 @@ impl SlidingWindow {
     }
 }
 
+/// Cached outcome, wether the item can be accepted, rejected or the cache has no information about
+/// this hash.
+#[derive(Debug)]
+enum CacheOutcome {
+    /// Hash accepted by cache.
+    Accepted,
+    /// Hash rejected by cache.
+    Rejected,
+    /// Cache has no information about the hash.
+    Unknown,
+}
+
+/// Internal cache remembering already accepted elements and current cardinality.
+///
+/// Only caches for the currently active granule of the sliding window.
+#[derive(Default)]
+struct Cache {
+    inner: RwLock<CacheInner>,
+}
+
+impl Cache {
+    /// Acquires a read lock from the cache and returns a read handle.
+    ///
+    /// All operations done on the handle share the same lock. To release the lock
+    /// the returned [`CacheRead`] must be dropped.
+    fn read(&self, organization_id: OrganizationId, window: u64, limit: usize) -> CacheRead<'_> {
+        let inner = self.inner.read().unwrap_or_else(PoisonError::into_inner);
+
+        // If the window does not match, we can already release the lock.
+        if window != inner.current_window {
+            return CacheRead::Empty;
+        }
+
+        CacheRead::Read {
+            inner,
+            organization_id,
+            limit,
+        }
+    }
+
+    /// Acquires a write lock from the cache and returns an update handle.
+    ///
+    /// All operations done on the handle share the same lock. To release the lock
+    /// the returned [`CacheUpdate`] must be dropped.
+    fn update<'a>(
+        &'a self,
+        organization_id: OrganizationId,
+        item_scope: &'a str,
+        window: u64,
+    ) -> CacheUpdate<'a> {
+        let mut inner = self.inner.write().unwrap_or_else(PoisonError::into_inner);
+
+        // If the passed window is newer then the current window, reset the cache to the new window.
+        if window > inner.current_window {
+            inner.current_window = window;
+            inner.cache.clear();
+        } else if window != inner.current_window {
```

Comment:
nit: I find this slightly easier to understand

```suggestion
        } else if window < inner.current_window {
```